source/Timer.h

```diff
@@ -0,0 +1,116 @@
+/* Timer.h
+Copyright (c) 2023 by Timothy Collett
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef ES_TIMER_H_
+#define ES_TIMER_H_
+
+#include "MissionAction.h"
+#include "System.h"
+
+#include <cstdint>
+#include <map>
+#include <string>
+
+class DataNode;
+class DataWriter;
+class Planet;
+class PlayerInfo;
+class UI;
+
+// Class representing a timer for triggering mission actions
+// This requires a specification of how long to wait (base + randint(rand)),
+// a system to wait in, and whether to reset the timer upon leaving the system
+// and can optionally require the player to be idle (no acceleration), and/or
+// within or beyond a certain proximity to the system center or a
+// specific (named) system object
+class Timer {
+public:
+	Timer() = default;
+	Timer(const DataNode &node, Mission *mission);
+	// Set up the timer from its data file node
+	void Load(const DataNode &node, Mission *mission);
+	// Note: the Save() function can assume this is an instantiated Timer, not a template,
+	// so the time to wait will be saved fully calculated, and with any elapsed time subtracted
+	void Save(DataWriter &out) const;
+
+	// Get the total time to wait, including the random value
+	uint64_t TimeToWait() const;
+	// Get whether the timer requires the player to be idle
+	bool RequireIdle() const;
+	// Get the required proximity radius to a system object or the system center
+	// If 0., no proximity is required
+	double Proximity() const;
+	// Get whether the player needs to be close to or far from the specified center
+	bool CloseTo() const;
+	// Get the system the timer is specified for
+	const System *GetSystem() const;
+	// Get the named system object used for calculating the proximity center
+	const Planet *ProximityCenter() const;
+	// Get whether the timer is currently active
+	bool IsComplete() const;
+
+	// Calculate the total time to wait, including any random value
+	Timer Instantiate(std::map<std::string, std::string> &subs,
+					const System *origin, int jumps, int64_t payload) const;
+	// Progress the timer within the main loop
+	void Step(PlayerInfo &player, UI *ui);
+
+private:
+	enum ResetCondition { NONE, PAUSE, LEAVE_ZONE, LEAVE_SYSTEM };
```

Comment:
Enum class please

Reply:
Again, my lack of C++ experience shows: is it as simple as adding `class` after the `enum`? Or is more required here?

Reply:
Yeah basically. You don't need to make the values all caps either

Reply:
...Is there a Style Guide on that last part? Because most, though not all, such `enum`s do currently seem to have all-caps names...